### DISCUSSION TITLE
Implement the ability to set button icon programmatically

### DISCUSSION THREAD

**Discussion by godlike64** [streamdeck-linux-gui/streamdeck-linux-gui#89] [Ideas / Feature Requests]
This is a continuation of [this issue](https://github.com/timothycrosley/streamdeck-ui/issues/99). The idea is to have a way to set the button icon automatically based on some condition. This is useful if you have a button that executes an action, and you need some kind of visual feedback to know the current state (the most common example is having a streamdeck button to mute/unmute your microphone, and have it show you the current microphone status).

Not saying this is the best way to go about it, but I implemented this with the following patch, which basically looks for the output of a command to see if it returned the path to an image, and sets the image in the button (my mute/unmute script figures out what will be the end state, and writes the path to the relevant image accordingly in its output):

~~~
$ diff gui.py.orig gui.py
7c7
< from subprocess import Popen  # nosec - Need to allow users to specify arbitrary commands
---
$ diff gui.py.orig gui.py
7c7
< from subprocess import Popen  # nosec - Need to allow users to specify arbitrary commands
---
> from subprocess import Popen, PIPE  # nosec - Need to allow users to specify arbitrary commands
8a9,10
> from pathlib import Path
> from select import select
150c152,159
<                 Popen(shlex.split(command))
---
>                 p = Popen(shlex.split(command), stdout=PIPE)
>                 #output = p.stdout.read()
>                 output = select([p.stdout], [], [], 0.1)[0]
>                 if len(output) > 0:
>                     output = output[0].read().decode('utf-8').replace('\n', '')
>                     image_path = Path(output)
>                     if image_path.exists() and not image_path.is_dir():
>                         api.set_button_icon(deck_id, page, key, str(image_path))
~~~

**Comment by coolapso** [comment]
I think this PR implements what you want: https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/pull/82 

you change the state of the button via CLI.  Or am i wrong @rkmax ?

**Reply by rkmax** [reply]
Yeah the cli tool allows you change the icons, text and other appearance of the bottom even without #82 which adds the functionality of create multi state button without the need of code
